core/src/main/java/org/apache/iceberg/avro/Avro.java

```diff
@@ -123,7 +127,7 @@ public WriteBuilder named(String newName) {
       return this;
     }
 
-    public WriteBuilder createWriterFunc(Function<Schema, DatumWriter<?>> writerFunction) {
+    public WriteBuilder createWriterFunc(Function<Schema, MetricsAwareDatumWriter<?>> writerFunction) {
```

Comment:
This is going to break existing uses of `createWriterFunc` in projects that build on Iceberg. I think this should keep the old parameter and just check whether the implementation is `MetricsAwareDatumWriter` in the appender to return metrics.

Reply:
Sounds good, I didn't think of the case where people have their own implementation of these interfaces so I totally missed this. Will update and keep in mind!